File: Tree/problem95/problem95.py
from typing import Dict, List, Optional, Tuple


# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def generateTrees(self, n: int) -> List[Optional["TreeNode"]]:
        # Memoize trees for each value range; clone left/right when forming a new root so
        # each answer tree owns its own nodes (no sharing across the result list).
        def clone(node: Optional["TreeNode"]) -> Optional["TreeNode"]:
            if node is None:
                return None
            return TreeNode(node.val, clone(node.left), clone(node.right))

        memo: Dict[Tuple[int, int], List[Optional["TreeNode"]]] = {}

        def gen(start: int, end: int) -> List[Optional["TreeNode"]]:
            if start > end:
                return [None]
            key = (start, end)
            if key in memo:
                return memo[key]
            out: List[Optional["TreeNode"]] = []
            for i in range(start, end + 1):
                for left in gen(start, i - 1):
                    for right in gen(i + 1, end):
                        out.append(TreeNode(i, clone(left), clone(right)))
            memo[key] = out
            return out

        return gen(1, n)

File: Tree/problem95/problem95.cpp
#include <functional>
#include <map>
#include <utility>
#include <vector>

using namespace std;

// LeetCode already provides TreeNode. For local compile: g++ -std=c++17 -DLOCAL_JUDGE -c problem95.cpp
#ifdef LOCAL_JUDGE
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
#endif

class Solution {
public:
    vector<TreeNode *> generateTrees(int n) {
        // Memoize trees for each value range; clone left/right when forming a new root so
        // each answer tree owns its own nodes (no sharing across the result list).
        function<TreeNode *(TreeNode *)> clone = [&](TreeNode *root) -> TreeNode * {
            if (!root) {
                return nullptr;
            }
            return new TreeNode(root->val, clone(root->left), clone(root->right));
        };

        map<pair<int, int>, vector<TreeNode *>> memo;

        function<vector<TreeNode *>(int, int)> gen = [&](int start, int end) -> vector<TreeNode *> {
            if (start > end) {
                return {nullptr};
            }
            pair<int, int> key = {start, end};
            auto it = memo.find(key);
            if (it != memo.end()) {
                return it->second;
            }
            vector<TreeNode *> res;
            for (int i = start; i <= end; ++i) {
                for (TreeNode *l : gen(start, i - 1)) {
                    for (TreeNode *r : gen(i + 1, end)) {
                        res.push_back(new TreeNode(i, clone(l), clone(r)));
                    }
                }
            }
            memo[key] = res;
            return res;
        };

        return gen(1, n);
    }
};
